app/web_conn.go

```diff
@@ -304,19 +414,117 @@ func (wc *WebConn) writePump() {
 	}
 }
 
-// writeMessage is a helper utility that wraps the write to the socket
+// writeMessageBuf is a helper utility that wraps the write to the socket
 // along with setting the write deadline.
-func (wc *WebConn) writeMessage(msgType int, data []byte) error {
+func (wc *WebConn) writeMessageBuf(msgType int, data []byte) error {
 	wc.WebSocket.SetWriteDeadline(time.Now().Add(writeWaitTime))
 	return wc.WebSocket.WriteMessage(msgType, data)
 }
 
+func (wc *WebConn) writeMessage(msg *model.WebSocketEvent) error {
+	// We don't use the encoder from the write pump because it's unwieldy to pass encoders
+	// around, and this is only called during initialization of the webConn.
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	err := msg.Encode(enc)
+	if err != nil {
+		mlog.Warn("Error in encoding websocket message", mlog.Err(err))
+		return nil
+	}
+	wc.Sequence++
+
+	return wc.writeMessageBuf(websocket.TextMessage, buf.Bytes())
+}
+
 // addToDeadQueue appends a message to the dead queue.
-func (wc *WebConn) addToDeadQueue(msg model.WebSocketMessage) {
+func (wc *WebConn) addToDeadQueue(msg *model.WebSocketEvent) {
 	wc.deadQueue[wc.deadQueuePointer] = msg
 	wc.deadQueuePointer = (wc.deadQueuePointer + 1) % deadQueueSize
 }
 
+// hasMsgLoss indicates whether the next wanted sequence is right after
+// the latest element in the dead queue, which would mean there is no message loss.
+func (wc *WebConn) hasMsgLoss() bool {
+	if wc.deadQueuePointer == 0 {
+		if wc.deadQueue[deadQueueSize-1] == nil {
+			return false // No msg written
+		}
+		if wc.deadQueue[deadQueueSize-1].GetSequence() == wc.Sequence-1 {
+			return false
+		}
+		return true
+	}
+	if wc.deadQueue[wc.deadQueuePointer-1].GetSequence() == wc.Sequence-1 {
+		return false
+	}
+	return true
+}
+
+// isInDeadQueue checks whether a given sequence number is in the dead queue or not.
+// And if it is, it returns that index.
+func (wc *WebConn) isInDeadQueue(seq int64) (bool, int) {
+	// Can be optimized to traverse backwards from deadQueuePointer
+	// Hopefully, traversing 128 elements is not too much overhead.
+	for i := 0; i < deadQueueSize; i++ {
+		elem := wc.deadQueue[i]
+		if elem == nil {
+			return false, 0
+		}
+
+		if elem.GetSequence() == seq {
+			return true, i
+		}
+	}
+	return false, 0
+}
+
+func (wc *WebConn) clearDeadQueue() {
+	for i := 0; i < deadQueueSize; i++ {
+		if wc.deadQueue[i] == nil {
+			return
+		}
+		wc.deadQueue[i] = nil
+	}
```

Comment:
If we are sure about `wc.deadQueuePointer`, could we use something like this?:
```suggestion
	copy(wc.deadQueue[:wc.deadQueuePointer], make([]*model.WebSocketEvent, wc.deadQueuePointer))
```

Reply:
We can, but that creates extra allocations. Actually, we can even do `wc.deadQueue = wc.deadQueue[:0]`. :wink: